### PR TITLE
feat: Detect orphaned manifests

### PR DESCRIPTION
**Pull request by Tomasz-Smelcerz-SAP**
**Description**

Implement orphaned manifest detection as described in: #2212 

Changes proposed in this pull request:

- Detection of orphaned Kyma manifest (Parent Kyma cannot be found)
- If orphaned, put the Manifest in the error state and do not process it
- emit a new metric: `manifest_orphaned`
- return an error from reconciliation

**Related issue(s)**
#2212

### CODE REVIEW

File: config/manager/kustomization.yaml [hunk 1]
```diff
@@ -3,9 +3,9 @@ kind: Kustomization
 generatorOptions:
   disableNameSuffixHash: true
 resources:
-  - manager.yaml
-  - metrics_service.yaml
+- manager.yaml
+- metrics_service.yaml
 images:
-  - name: controller
```
Comment (LeelaChacha) [inline config/manager/kustomization.yaml:9]: please remove from final PR

File: internal/service/manifest/orphan/detection.go [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+	"errors"
+	"fmt"
+
+	"github.com/kyma-project/lifecycle-manager/api/v1beta2"
+	"github.com/kyma-project/lifecycle-manager/pkg/util"
+)
+
+var ErrOrphanedManifest = errors.New("orphaned manifest detected")
+
+type APIClient interface {
```
Comment (lindnerby) [inline internal/service/manifest/orphan/detection.go:14]: Rename this to Repository please.

File: unit-test-coverage.yaml [hunk 3]
```diff
@@ -19,9 +19,11 @@ packages:
   internal/maintenancewindows: 87
   internal/pkg/resources: 91
   internal/remote: 29
+  internal/repository/manifest/orphan: 100
   internal/service/kyma/status/modules: 78
   internal/service/kyma/status/modules/generator: 100
   internal/service/kyma/status/modules/generator/fromerror: 100
+  internal/service/manifest/orphan: 95.8
```
Comment (lindnerby) [inline unit-test-coverage.yaml:26]: Let's push this to 💯 !

File: internal/service/manifest/orphan/detection.go [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+func (s *DetectionService) DetectOrphanedManifest(ctx context.Context, manifest *v1beta2.Manifest) error {
+	if manifest.IsMandatoryModule() {
+		// Mandatory modules are not refereced by any Kyma object so cannot be orphaned
+		return nil
+	}
+
+	if manifest.GetDeletionTimestamp() != nil {
+		// If the manifest is being deleted, we don't check for orphaned status (as it should be eventually deleted)
+		return nil
+	}
```
Comment (lindnerby) [inline internal/service/manifest/orphan/detection.go:37]: Can you wrap all the skip cases into one func please

File: internal/declarative/v2/reconciler.go [hunk 5]
```diff
@@ -133,6 +141,16 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
 		return ctrl.Result{RequeueAfter: r.Success}, nil
 	}
 
+	err = r.orphanDetectionService.DetectOrphanedManifest(ctx, manifest)
+	if err != nil {
+		if errors.Is(err, orphan.ErrOrphanedManifest) && !isManifestRecentlyCreated(manifest.CreationTimestamp.Time) {
```
Comment (lindnerby) [inline internal/declarative/v2/reconciler.go:146]: This can also be part of the detection service!!

So you can even skip the cyclomatic complexity linter suppression ;)

File: internal/declarative/v2/reconciler.go [hunk 6]
```diff
@@ -210,6 +228,10 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
 	return r.finishReconcile(ctx, manifest, metrics.ManifestReconcileFinished, manifestStatus, nil)
 }
 
+func isManifestRecentlyCreated(manifestCreationTimestamp time.Time) bool {
```
Comment (lindnerby) [inline internal/declarative/v2/reconciler.go:231]: Move to service as well

File: internal/repository/manifest/orphan/client.go [hunk 7]
```diff
@@ -0,0 +1,29 @@
+package orphan
```
Comment (lindnerby) [inline internal/repository/manifest/orphan/client.go:1]: The repository impl does not know of the use case it is used in "orphaned manifest etc."
You can just put this repo impl to `internal/repository/kyma/client.go`, since it is actually a kyma repository implementation.

File: internal/service/manifest/orphan/detection_test.go [hunk 8]
```diff
@@ -0,0 +1,245 @@
+package orphan_test
```
Comment (lindnerby) [inline internal/service/manifest/orphan/detection_test.go:1]: To reach 100% Copilot is suggesting to add:

1. Error when manifest.GetKymaName() fails in getParentKyma: Add a test case where manifest.GetKymaName() returns an error. This will cover the error path in the getParentKyma method.
Error when util.IsNotFound(err) is false in getParentKyma:


2. Add a test case where s.client.GetKyma returns an error that is not a "not found" error, and ensure the correct error is returned.
Edge case for isManifestReferencedInKymaStatus:


3. Add a test case where kyma.Status.Modules contains a nil entry to ensure the function handles it gracefully.

File: internal/declarative/v2/reconciler.go (hunk 5 above)
Comment (lindnerby) [reply internal/declarative/v2/reconciler.go:146]: Basically everything except setting the status should be part of the service.

File: internal/service/manifest/orphan/detection_test.go [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+
+	err := service.DetectOrphanedManifest(t.Context(), manifest)
+
+	require.Error(t, err)
+	require.ErrorIs(t, err, orphan.ErrOrphanedManifest)
+}
+
+func TestDetectionService_DetectOrphanedManifest_WhenClientReturnsError_ReturnError(t *testing.T) {
+	service := orphan.NewDetectionService(&clientStubGenericError{})
+	manifest := &v1beta2.Manifest{
```
Comment (lindnerby) [inline internal/service/manifest/orphan/detection_test.go:76]: For these cases where we just want to create a default manifest with kyma reference, maybe a test helper func?

File: internal/pkg/metrics/manifest.go [hunk 10]
```diff
@@ -33,6 +33,7 @@ const (
 	ManifestReconcileFinished            ManifestRequeueReason = "manifest_reconcile_finished"
 	ManifestUnmanagedUpdate              ManifestRequeueReason = "manifest_unmanaged_update"
 	ManifestResourcesLabelRemoval        ManifestRequeueReason = "manifest_labels_removal"
+	ManifestOrphaned                     ManifestRequeueReason = "manifest_orphaned"
```
Comment (lindnerby) [inline internal/pkg/metrics/manifest.go:36]: Since we were talking about the cleanup yesterday:

Just setting this metric should be fine, we are calling cleanup from the metrics pkg:
`internal/pkg/metrics/manifest.go`

**Review by lindnerby — CHANGES_REQUESTED** [review]

File: internal/service/manifest/orphan/detection_test.go (hunk 8 above)
Comment (LeelaChacha) [reply internal/service/manifest/orphan/detection_test.go:1]: 2. is already accounted for in test "TestDetectionService_DetectOrphanedManifest_WhenClientReturnsError_ReturnError"
3. is already accounted for in "TestDetectionService_DetectOrphanedManifest_WhenModuleStatusManifestNil_ReturnOrphanedManifestError"

File: internal/declarative/v2/reconciler.go (hunk 5 above)
Comment (LeelaChacha) [reply internal/declarative/v2/reconciler.go]: cyclomatic complexity existed before the check for manifest creation timestamp. other parts of the function will have to be refined to get the complexity down, since the if block for the error is essential to setting the status

**Review by lindnerby — APPROVED** [review]
